src/compiler/rewriter.ts

```diff
@@ -0,0 +1,1505 @@
+                }
+            }
+            else if (node.operator === SyntaxKind.CommaToken) {
+                return rewriteCommaExpression(node, state);
+            }
+            else if (hasAwaitOrYield(node.right)) {
+                return Factory.updateBinaryExpression(
+                    node,
+                    cacheExpression(Visitor.visit(node.left, visitNode, state), state),
+                    Visitor.visit(node.right, visitNode, state));
+            }
+            return Visitor.fallback(node, visitNode, state);
+        }
+
+        function rewriteLogicalBinaryExpression(node: BinaryExpression, state: RewriterState): Expression {
+            var builder = state.builder;
+            var resumeLabel = GeneratorFunctionBuilder.defineLabel(builder);
+            var resultLocal = GeneratorFunctionBuilder.declareLocal(builder);
+            var code = node.operator === SyntaxKind.AmpersandAmpersandToken ? OpCode.BreakWhenFalse : OpCode.BreakWhenTrue;
+            GeneratorFunctionBuilder.writeLocation(builder, node.left);
+            GeneratorFunctionBuilder.emit(builder, OpCode.Assign, resultLocal, Visitor.visit(node.left, visitNode, state));
+            GeneratorFunctionBuilder.emit(builder, code, resumeLabel, resultLocal);
+            GeneratorFunctionBuilder.writeLocation(builder, node.right);
+            GeneratorFunctionBuilder.emit(builder, OpCode.Assign, resultLocal, Visitor.visit(node.right, visitNode, state));
+            GeneratorFunctionBuilder.markLabel(builder, resumeLabel);
+            return resultLocal;
+        }
+
+        function rewriteCommaExpression(node: BinaryExpression, state: RewriterState): Expression {
+            var builder = state.builder;
+            var expressions = flattenCommaExpression(node);
+            var merged: Expression;
+            for (var i = 0; i < expressions.length; i++) {
+                var expression = expressions[i];
+                if (hasAwaitOrYield(expression) && merged) {
+                    GeneratorFunctionBuilder.emit(builder, OpCode.Statement, Factory.createExpressionStatement(merged));
+                    merged = undefined;
+                }
+                var visited = Visitor.visit(expression, visitNode, state);
+                if (merged) {
+                    merged = Factory.createBinaryExpression(
+                        SyntaxKind.CommaToken,
+                        merged,
+                        visited);
+                }
+                else {
+                    merged = visited;
+                }
+            }
+            return merged;
+        }
+
+        function rewriteDestructuringAssignment(node: BinaryExpression, state: RewriterState): Expression {
+            var destructured = DestructuringAssignmentRewriter.rewrite(node, state.locals);
+            var rewritten = visitBinaryExpression(destructured, state);
+            if (needsParenthesisForPropertyAccess(node)) {
```

Comment:
I've suggested that a PRer should rename this as `needsParenthesesForPropertyAccessOrInvocation` so be aware for the future.